File: scripts/hub/rule_processor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
统一的规则处理器 - 消除 ruleset_manager 和 adblock_manager 之间的重复代码
提供规则标准化、去重、验证的单一来源
"""

import re
from typing import Optional, Set, Dict, List

from hub.surge_compliance import (
    INVALID_DOMAIN_REGEX_VALUES,
    INVALID_URL_REGEX_VALUES,
    is_invalid_domain_regex_payload,
    is_invalid_url_regex_payload,
    strip_inline_comment,
    format_url_regex_for_surge,
)

# 规则类型正则
DOMAIN_RULE = re.compile(r'^DOMAIN,([^,]+)$', re.IGNORECASE)
DOMAIN_SUFFIX_RULE = re.compile(r'^DOMAIN-SUFFIX,([^,]+)$', re.IGNORECASE)
DOMAIN_KEYWORD_RULE = re.compile(r'^DOMAIN-KEYWORD,([^,]+)$', re.IGNORECASE)
IP_CIDR_RULE = re.compile(r'^IP-CIDR6?,([^,]+)$', re.IGNORECASE)
DOMAIN_REGEX_RULE = re.compile(r'^DOMAIN-REGEX,(.+)$', re.IGNORECASE)

# 域名验证（宽松模式，允许通配符）
VALID_DOMAIN = re.compile(
    r'^(\*\.)?[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?'
    r'(\.[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$'
)

# IP-CIDR 验证
VALID_IPV4_CIDR = re.compile(r'^(\d{1,3}\.){3}\d{1,3}(/\d{1,2})?$')
VALID_IPV6_CIDR = re.compile(r'^([0-9a-fA-F:]+)(/\d{1,3})?$')


class RuleProcessor:
    """统一的规则处理器"""

    def __init__(self, strict_validation: bool = False):
        """
        Args:
            strict_validation: 是否启用严格验证（拒绝可疑规则）
        """
        self.strict_validation = strict_validation
        self.stats = {
            'total': 0,
            'normalized': 0,
            'invalid': 0,
            'duplicate': 0,
        }

    def normalize_rule(self, line: str, source: str = "") -> Optional[str]:
        """标准化单条规则

        Args:
            line: 原始规则行
            source: 规则来源（用于日志）

        Returns:
            标准化后的规则，无效返回 None
        """
        self.stats['total'] += 1

        line = strip_inline_comment(line.strip())
        if not line or line.startswith(('#', ';')):
            return None

        if ',' not in line:
            self.stats['invalid'] += 1
            return None

        rule_type, rule_value = line.split(',', 1)
        rule_type = rule_type.strip().upper()
        rule_value = rule_value.strip()

        if rule_type.startswith("IP-"):
            rule_value = rule_value.split(",", 1)[0].strip()

        if rule_type not in ('DOMAIN', 'DOMAIN-SUFFIX', 'DOMAIN-KEYWORD',
                              'IP-CIDR', 'IP-CIDR6', 'DOMAIN-REGEX', 'GEOIP',
                              'USER-AGENT', 'URL-REGEX', 'PROCESS-NAME'):
            self.stats['invalid'] += 1
            return None

        if rule_value.startswith('"') and rule_value.endswith('"'):
            rule_value = rule_value[1:-1]

        if not self._validate_rule_value(rule_type, rule_value):
            self.stats['invalid'] += 1
            return None

        normalized = self._format_rule(rule_type, rule_value)
        self.stats['normalized'] += 1
        return normalized

    @staticmethod
    def _format_rule(rule_type: str, rule_value: str) -> str:
        if rule_type == "URL-REGEX":
            return format_url_regex_for_surge(rule_value)
        return f"{rule_type},{rule_value}"

    def _validate_rule_value(self, rule_type: str, value: str) -> bool:
        """验证规则值的有效性"""
        if not value:
            return False

        if rule_type in ('DOMAIN', 'DOMAIN-SUFFIX'):
            value = value.lstrip('.')
            if self.strict_validation:
                return bool(VALID_DOMAIN.match(value))
            return '.' in value or len(value) > 2

        if rule_type == 'DOMAIN-KEYWORD':
            return len(value) >= 2 and not any(c in value for c in '<>{}()')

        if rule_type == 'IP-CIDR':
            return bool(VALID_IPV4_CIDR.match(value))
        if rule_type == 'IP-CIDR6':
            return bool(VALID_IPV6_CIDR.match(value))

        if rule_type == 'DOMAIN-REGEX':
            if is_invalid_domain_regex_payload(value):
                return False
            try:
                re.compile(value)
                return True
            except re.error:
                return False

        if rule_type == 'URL-REGEX':
            if is_invalid_url_regex_payload(value):
                return False
            try:
                re.compile(value)
                return True
            except re.error:
                return False

        return True

    def deduplicate_rules(self, rules: List[str]) -> List[str]:
        seen = set()
        result = []
        for rule in rules:
            if rule not in seen:
                seen.add(rule)
                result.append(rule)
            else:
                self.stats['duplicate'] += 1
        return result

    def deduplicate_by_priority(self, rules_dict: Dict[str, Set[str]]) -> Dict[str, Set[str]]:
        rule_to_sources: Dict[str, List[str]] = {}
        for source, rules in rules_dict.items():
            for rule in rules:
                if rule not in rule_to_sources:
                    rule_to_sources[rule] = []
                rule_to_sources[rule].append(source)

        result = {source: set() for source in rules_dict.keys()}
        for rule, sources in rule_to_sources.items():
            if len(sources) == 1:
                result[sources[0]].add(rule)
            else:
                primary = sorted(sources)[0]
                result[primary].add(rule)
                self.stats['duplicate'] += len(sources) - 1

        return result

    def extract_domains_from_rules(self, rules: List[str]) -> Set[str]:
        domains = set()
        for rule in rules:
            m = DOMAIN_RULE.match(rule)
            if m:
                domains.add(m.group(1))
                continue
            m = DOMAIN_SUFFIX_RULE.match(rule)
            if m:
                domains.add(m.group(1))
                continue
        return domains

    def filter_by_whitelist(self, rules: List[str], whitelist: Set[str]) -> List[str]:
        result = []
        for rule in rules:
            m = DOMAIN_RULE.match(rule) or DOMAIN_SUFFIX_RULE.match(rule)
            if m:
                domain = m.group(1).lower()
                if domain in whitelist:
                    continue
                if any(domain.endswith('.' + w) for w in whitelist):
                    continue
            result.append(rule)
        return result

    def split_by_type(self, rules: List[str]) -> Dict[str, List[str]]:
        result: Dict[str, List[str]] = {}
        for rule in rules:
            parts = rule.split(',', 1)
            if len(parts) >= 2:
                rule_type = parts[0].upper()
                if rule_type not in result:
                    result[rule_type] = []
                result[rule_type].append(rule)
        return result

    def get_stats(self) -> Dict[str, int]:
        return self.stats.copy()

    def reset_stats(self):
        self.stats = {
            'total': 0,
            'normalized': 0,
            'invalid': 0,
            'duplicate': 0,
        }


_default_processor = RuleProcessor()


def normalize_rule(line: str, source: str = "") -> Optional[str]:
    return _default_processor.normalize_rule(line, source)


def deduplicate_rules(rules: List[str]) -> List[str]:
    return _default_processor.deduplicate_rules(rules)

File: create/hub/rule_processor.go
package hub

import (
	"regexp"
	"sort"
	"strings"
)

var surgePolicyOptions = map[string]bool{
	"DIRECT": true, "REJECT": true, "REJECT-DROP": true, "REJECT-NO-DROP": true,
	"REJECT-TINYGIF": true, "REJECT-IMG": true, "PROXY": true,
}

var (
	domainRule        = regexp.MustCompile(`(?i)^DOMAIN,([^,]+)$`)
	domainSuffixRule  = regexp.MustCompile(`(?i)^DOMAIN-SUFFIX,([^,]+)$`)
	domainKeywordRule = regexp.MustCompile(`(?i)^DOMAIN-KEYWORD,([^,]+)$`)
	ipCidrRule        = regexp.MustCompile(`(?i)^IP-CIDR6?,([^,]+)$`)
	domainRegexRule   = regexp.MustCompile(`(?i)^DOMAIN-REGEX,(.+)$`)
)

var (
	validDomain   = regexp.MustCompile(`^(\*\.)?[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(\.[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$`)
	validIpv4Cidr = regexp.MustCompile(`^(\d{1,3}\.){3}\d{1,3}(/\d{1,2})?$`)
	validIpv6Cidr = regexp.MustCompile(`^([0-9a-fA-F:]+)(/\d{1,3})?$`)
)

type RuleProcessor struct {
	strictValidation bool
	stats            map[string]int
}

func NewRuleProcessor(strictValidation bool) *RuleProcessor {
	return &RuleProcessor{
		strictValidation: strictValidation,
		stats: map[string]int{
			"total":      0,
			"normalized": 0,
			"invalid":    0,
			"duplicate":  0,
		},
	}
}

func (rp *RuleProcessor) NormalizeRule(line string, source string) *string {
	rp.stats["total"]++

	line = StripInlineComment(strings.TrimSpace(line))
	if line == "" || strings.HasPrefix(line, "#") || strings.HasPrefix(line, ";") {
		return nil
	}

	if !strings.Contains(line, ",") {
		rp.stats["invalid"]++
		return nil
	}

	parts := strings.SplitN(line, ",", 2)
	ruleType := strings.ToUpper(strings.TrimSpace(parts[0]))
	ruleValue := strings.TrimSpace(parts[1])

	if strings.HasPrefix(ruleType, "IP-") {
		ruleValue = strings.TrimSpace(strings.SplitN(ruleValue, ",", 2)[0])
	}

	validTypes := map[string]bool{
		"DOMAIN": true, "DOMAIN-SUFFIX": true, "DOMAIN-KEYWORD": true,
		"IP-CIDR": true, "IP-CIDR6": true, "DOMAIN-REGEX": true, "GEOIP": true,
		"USER-AGENT": true, "URL-REGEX": true, "PROCESS-NAME": true,
	}
	if !validTypes[ruleType] {
		rp.stats["invalid"]++
		return nil
	}

	if ruleType == "URL-REGEX" {
		// External RULE-SET entries carry the pattern only; strip a source policy
		// suffix before formatting so quoted patterns are not double-wrapped.
		if strings.HasPrefix(ruleValue, "\"\"") {
			ruleValue = ruleValue[1:]
		}
		body, options, validShape := splitURLRegexPayload(ruleValue)
		if !validShape || len(body) == 0 {
			rp.stats["invalid"]++
			return nil
		}
		for _, option := range options {
			option = strings.ToUpper(strings.TrimSpace(option))
			if option != "" && option != "EXTENDED-MATCHING" && !surgePolicyOptions[option] {
				rp.stats["invalid"]++
				return nil
			}
		}
		ruleValue = body
	} else if strings.HasPrefix(ruleValue, "\"") && strings.HasSuffix(ruleValue, "\"") && len(ruleValue) >= 2 {
		ruleValue = ruleValue[1 : len(ruleValue)-1]
	}

	if !rp.validateRuleValue(ruleType, ruleValue) {
		rp.stats["invalid"]++
		return nil
	}

	normalized := formatRule(ruleType, ruleValue)
	rp.stats["normalized"]++
	return &normalized
}

func formatRule(ruleType string, ruleValue string) string {
	if ruleType == "URL-REGEX" {
		return FormatUrlRegexForSurge(ruleValue)
	}
	return ruleType + "," + ruleValue
}

func (rp *RuleProcessor) validateRuleValue(ruleType string, value string) bool {
	if value == "" {
		return false
	}

	if ruleType == "DOMAIN" || ruleType == "DOMAIN-SUFFIX" {
		value = strings.TrimLeft(value, ".")
		if rp.strictValidation {
			return validDomain.MatchString(value)
		}
		return strings.Contains(value, ".") || len(value) > 2
	}

	if ruleType == "DOMAIN-KEYWORD" {
		if len(value) < 2 {
			return false
		}
		for _, c := range []rune{'<', '>', '{', '}', '(', ')'} {
			if strings.ContainsRune(value, c) {
				return false
			}
		}
		return true
	}

	if ruleType == "IP-CIDR" {
		return validIpv4Cidr.MatchString(value)
	}
	if ruleType == "IP-CIDR6" {
		return validIpv6Cidr.MatchString(value)
	}

	if ruleType == "DOMAIN-REGEX" {
		if IsInvalidDomainRegexPayload(value) {
			return false
		}
		_, err := regexp.Compile(value)
		return err == nil
	}

	if ruleType == "URL-REGEX" {
		if IsInvalidUrlRegexPayload(value) {
			return false
		}
		_, err := regexp.Compile(value)
		return err == nil
	}

	return true
}

func (rp *RuleProcessor) DeduplicateRules(rules []string) []string {
	seen := make(map[string]bool)
	var result []string
	for _, rule := range rules {
		if !seen[rule] {
			seen[rule] = true
			result = append(result, rule)
		} else {
			rp.stats["duplicate"]++
		}
	}
	return result
}

func (rp *RuleProcessor) DeduplicateByPriority(rulesDict map[string][]string) map[string][]string {
	ruleToSources := make(map[string][]string)
	for source, rules := range rulesDict {
		for _, rule := range rules {
			ruleToSources[rule] = append(ruleToSources[rule], source)
		}
	}

	result := make(map[string][]string)
	for source := range rulesDict {
		result[source] = []string{} // Initialize all sources
	}

	// Ensure stable deduplication by sorting rules and sources
	var sortedRules []string
	for rule := range ruleToSources {
		sortedRules = append(sortedRules, rule)
	}
	sort.Strings(sortedRules)

	for _, rule := range sortedRules {
		sources := ruleToSources[rule]
		if len(sources) == 1 {
			result[sources[0]] = append(result[sources[0]], rule)
		} else {
			sort.Strings(sources)
			primary := sources[0]
			result[primary] = append(result[primary], rule)
			rp.stats["duplicate"] += len(sources) - 1
		}
	}

	return result
}

func (rp *RuleProcessor) ExtractDomainsFromRules(rules []string) []string {
	domainsSet := make(map[string]bool)
	for _, rule := range rules {
		if m := domainRule.FindStringSubmatch(rule); m != nil {
			domainsSet[m[1]] = true
			continue
		}
		if m := domainSuffixRule.FindStringSubmatch(rule); m != nil {
			domainsSet[m[1]] = true
			continue
		}
	}
	var domains []string
	for d := range domainsSet {
		domains = append(domains, d)
	}
	sort.Strings(domains)
	return domains
}

func (rp *RuleProcessor) FilterByWhitelist(rules []string, whitelist map[string]bool) []string {
	var result []string
	for _, rule := range rules {
		var domain string
		if m := domainRule.FindStringSubmatch(rule); m != nil {
			domain = strings.ToLower(m[1])
		} else if m := domainSuffixRule.FindStringSubmatch(rule); m != nil {
			domain = strings.ToLower(m[1])
		}

		if domain != "" {
			if whitelist[domain] {
				continue
			}
			skip := false
			for w := range whitelist {
				if strings.HasSuffix(domain, "."+w) {
					skip = true
					break
				}
			}
			if skip {
				continue
			}
		}
		result = append(result, rule)
	}
	return result
}

func (rp *RuleProcessor) SplitByType(rules []string) map[string][]string {
	result := make(map[string][]string)
	for _, rule := range rules {
		parts := strings.SplitN(rule, ",", 2)
		if len(parts) >= 2 {
			ruleType := strings.ToUpper(strings.TrimSpace(parts[0]))
			result[ruleType] = append(result[ruleType], rule)
		}
	}
	return result
}

func (rp *RuleProcessor) GetStats() map[string]int {
	res := make(map[string]int)
	for k, v := range rp.stats {
		res[k] = v
	}
	return res
}

func (rp *RuleProcessor) ResetStats() {
	rp.stats = map[string]int{
		"total":      0,
		"normalized": 0,
		"invalid":    0,
		"duplicate":  0,
	}
}

var defaultProcessor = NewRuleProcessor(false)

func NormalizeRule(line string, source string) *string {
	return defaultProcessor.NormalizeRule(line, source)
}

func DeduplicateRules(rules []string) []string {
	return defaultProcessor.DeduplicateRules(rules)
}
